core/include/traccc/clusterization/details/sparse_ccl.hpp

```diff
@@ -47,8 +47,9 @@ TRACCC_HOST_DEVICE inline unsigned int make_union(
 ///
 /// @return boolan to indicate 8-cell connectivity
 ///
-TRACCC_HOST_DEVICE inline bool is_adjacent(const traccc::cell& a,
-                                           const traccc::cell& b);
+template <typename T1, typename T2>
+TRACCC_HOST_DEVICE inline bool is_adjacent(const edm::silicon_cell<T1>& a,
+                                           const edm::silicon_cell<T2>& b);
```

Comment:
I'd advocate for simply having two universal reference arguments, the `const edm::silicon_cell<T1>` syntax is highly confusing; it's not clear at all to me what `T1` and `T2` are even supposed to be.

It would be better to implement a concept for silicon cells and use that instead:

```suggestion
template <silicon_cell_like T1, silicon_cell_like T2>
TRACCC_HOST_DEVICE inline bool is_adjacent(T1 && a, T2 && b);
```

Reply:
I'm also okay with some typedefs thrown around, e.g. having some typedef for host cells and one for device cells..

Reply:
I really don't agree on this one. You want to wrote even more code, when the C\+\+11 templating I use here is already forcing the user to use a specific type of class in the argument list. :thinking:

I did not spell it out in the PR, but the separate `T1` and `T2` template arguments are needed to allow const and non-const, device and host cell proxies to be used in a single function, in any combination.

Honestly, you should get used to this syntax, as I feel pretty strongly about it being the correct way to go. :thinking:

Reply:
Okay, I can't say I agree but it's surely nor worth arguing over. :smile: